File: issuer-list-ui/src/main/java/com/adyen/checkout/issuerlist/IssuerListSpinnerAdapter.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 26/4/2019.
 */

package com.adyen.checkout.issuerlist;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import com.adyen.checkout.base.api.ImageLoader;
import com.adyen.checkout.issuerlist.ui.R;

import java.util.List;

public class IssuerListSpinnerAdapter extends BaseAdapter {

    private final LayoutInflater mInflater;

    private List<IssuerModel> mIssuerList;

    private final boolean mHideIssuerLogo;
    private final ImageLoader mImageLoader;
    private final String mPaymentMethod;

    IssuerListSpinnerAdapter(@NonNull Context context, @NonNull List<IssuerModel> issuerList, ImageLoader imageLoader, String paymentMethod,
            boolean hideIssuerLogo) {
        mInflater = LayoutInflater.from(context);
        mIssuerList = issuerList;
        mHideIssuerLogo = hideIssuerLogo;
        mImageLoader = imageLoader;
        mPaymentMethod = paymentMethod;
    }

    void updateIssuers(@NonNull List<IssuerModel> issuerList) {
        mIssuerList = issuerList;
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return mIssuerList.size();
    }

    @Override
    @NonNull
    public IssuerModel getItem(int position) {
        return mIssuerList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    @NonNull
    public View getView(int position, @Nullable View convertView, @Nullable ViewGroup parent) {
        final View view;
        final IssuerModel issuer = getItem(position);

        // TODO: 15/03/2019 check if optimization with ViewHolder is possible
        if (convertView != null) {
            view = convertView;
        } else {
            view = mInflater.inflate(R.layout.spinner_list_with_image, parent, false);
        }

        final AppCompatImageView issuerLogo = view.findViewById(R.id.imageView_logo);
        final AppCompatTextView issuerName = view.findViewById(R.id.textView_text);

        issuerName.setText(issuer.getName());
        if (!mHideIssuerLogo) {
            mImageLoader.load(mPaymentMethod,
                    issuer.getId(),
                    issuerLogo,
                    R.drawable.ic_placeholder_image,
                    R.drawable.ic_placeholder_image);
        } else {
            issuerLogo.setVisibility(View.GONE);
        }

        return view;
    }
}

File: issuer-list/src/main/java/com/adyen/checkout/issuerlist/IssuerListSpinnerAdapter.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 26/4/2019.
 */
package com.adyen.checkout.issuerlist

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.adyen.checkout.components.api.ImageLoader
import com.adyen.checkout.components.ui.databinding.SpinnerListWithImageBinding

class IssuerListSpinnerAdapter internal constructor(
    private val context: Context,
    private var issuerList: List<IssuerModel>,
    private val imageLoader: ImageLoader,
    private val paymentMethod: String,
    private val hideIssuerLogo: Boolean
) : BaseAdapter() {

    override fun getCount(): Int {
        return issuerList.size
    }

    override fun getItem(position: Int): IssuerModel {
        return issuerList[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        val view: View
        val viewHolder: IssuerListSpinnerViewHolder
        val binding: SpinnerListWithImageBinding
        if (convertView == null) {
            binding = SpinnerListWithImageBinding.inflate(LayoutInflater.from(context), parent, false)
            view = binding.root
            viewHolder = IssuerListSpinnerViewHolder(binding, imageLoader, paymentMethod, hideIssuerLogo)
        } else {
            view = convertView
            viewHolder = view.tag as IssuerListSpinnerViewHolder
        }
        viewHolder.bind(getItem(position))
        return view
    }
}

internal class IssuerListSpinnerViewHolder(
    private val binding: SpinnerListWithImageBinding,
    private val imageLoader: ImageLoader,
    private val paymentMethod: String,
    private val hideIssuerLogo: Boolean
) : ViewHolder(binding.root) {

    fun bind(model: IssuerModel) {
        binding.textViewText.text = model.name
        if (!hideIssuerLogo) {
            imageLoader.load(
                paymentMethod,
                model.id,
                binding.imageViewLogo,
                R.drawable.ic_placeholder_image,
                R.drawable.ic_placeholder_image
            )
        } else {
            binding.imageViewLogo.isVisible = false
        }
    }

}
